### ISSUE TITLE
Include :: IndexOutOfRange when running include queries with one-to-many and one-to-many self reference

### ISSUE THREAD

**Issue by maumar** [aspnet/EntityFramework#1433]
Repro:

```
    public class Level1
    {
        public int Id { get; set; }
        public string Name { get; set; }

        public List<Level2> OneToMany_Optional { get; set; }
        public List<Level1> OneToMany_Optional_Self { get; set; }
        public Level1 OneToMany_Optional_Self_Inverse { get; set; }
    }

    public class Level2
    {
        public int Id { get; set; }
        public string Name { get; set; }

        public int? Level1_Optional_Id { get; set; }
        public Level1 OneToMany_Optional_Inverse { get; set; }
    }

    public class MyContext : DbContext
    {
        public DbSet<Level1> LevelOne { get; set; }
        public DbSet<Level2> LevelTwo { get; set; }

        protected override void OnModelCreating(ModelBuilder modelBuilder)
        {
            base.OnModelCreating(modelBuilder);

            modelBuilder.Entity<Level1>().Property(e => e.Id).GenerateValueOnAdd(false);
            modelBuilder.Entity<Level2>().Property(e => e.Id).GenerateValueOnAdd(false);
            modelBuilder.Entity<Level1>().OneToMany(e => e.OneToMany_Optional, e => e.OneToMany_Optional_Inverse).Required(false);
            modelBuilder.Entity<Level1>().OneToMany(e => e.OneToMany_Optional_Self, e => e.OneToMany_Optional_Self_Inverse).Required(false);
            modelBuilder.Entity<Level2>().Property(e => e.Id).GenerateValueOnAdd(false);
            modelBuilder.Entity<Level2>().Property(e => e.Id).GenerateValueOnAdd(false);
        }

        protected override void OnConfiguring(DbContextOptions options)
        {
            options.UseSqlServer(new SqlConnectionStringBuilder { DataSource = ".", InitialCatalog = "IncludeBug_IndexOutsideBounds", MultipleActiveResultSets = true, IntegratedSecurity = true }.ToString());
        }
    }

    class Program
    {
        static void Main(string[] args)
        {
            using (var ctx = new MyContext())
            {
                ctx.Database.EnsureDeleted();
                ctx.Database.EnsureCreated();

                var l1_01 = new Level1 { Id = 1, Name = "L1 01", };
                var l1_02 = new Level1 { Id = 2, Name = "L1 02", };
                var l1_03 = new Level1 { Id = 3, Name = "L1 03", };

                var l2_01 = new Level2 { Id = 1, Name = "L2 01", };
                var l2_02 = new Level2 { Id = 2, Name = "L2 02", };
                var l2_03 = new Level2 { Id = 3, Name = "L2 03", };

                ctx.LevelOne.Add(l1_01, l1_02, l1_03);
                ctx.LevelTwo.Add(l2_01, l2_02, l2_03);

                l1_01.OneToMany_Optional = new List<Level2> { l2_02, };
                l1_02.OneToMany_Optional_Self = new List<Level1> { l1_01 };

                ctx.SaveChanges();
            }

            using (var ctx = new MyContext())
            {
                ctx.LevelOne.Include(e => e.OneToMany_Optional).ToList();
                ctx.LevelOne.Include(e => e.OneToMany_Optional_Self).ToList();
            }
        }
    }

```

this throws:

Unhandled Exception: System.ArgumentOutOfRangeException: Index was out of range.
 Must be non-negative and less than the size of the collection.
Parameter name: index
   at System.ThrowHelper.ThrowArgumentOutOfRangeException(ExceptionArgument argument, ExceptionResource resource)
   at System.Collections.Generic.List`1.get_Item(Int32 index)
   at Microsoft.Data.Entity.Query.QueryBuffer.GetTargetEntity(IEntityType targetEntityType, IValueReader valueReader, ICollection`1 bufferedEntities) in D:\k\EntityFramework\src\EntityFramework.Core\Query\QueryBuffer.cs:line 401
   at Microsoft.Data.Entity.Query.QueryBuffer.<>c__DisplayClass0.<Include>b__2(IValueReader valueReader) in D:\k\EntityFramework\src\EntityFramework.Core\Query\QueryBuffer.cs:line 195
   at System.Linq.Enumerable.WhereSelectEnumerableIterator`2.MoveNext()
   at System.Linq.Enumerable.WhereEnumerableIterator`1.MoveNext()
   at System.Collections.Generic.List`1..ctor(IEnumerable`1 collection)
   at System.Linq.Enumerable.ToList[TSource](IEnumerable`1 source)
   at Microsoft.Data.Entity.Query.QueryBuffer.Include(Object entity, IReadOnlyList`1 navigationPath, IReadOnlyList`1 relatedValueReaders, Int32 currentNavigationIndex) in D:\k\EntityFramework\src\EntityFramework.Core\Query\QueryBuffer.cs:line 184
   at Microsoft.Data.Entity.Query.QueryBuffer.Include(Object entity, IReadOnlyList`1 navigationPath, IReadOnlyList`1 relatedValueReaders) in D:\k\EntityFramework\src\EntityFramework.Core\Query\QueryBuffer.cs:line 154
   at Microsoft.Data.Entity.Relational.Query.QueryMethodProvider.<_IncludeCollection>d__0`1.MoveNext() in D:\k\EntityFramework\src\EntityFramework.Relational\Qu
ery\QueryMethodProvider.cs:line 58
   at Microsoft.Data.Entity.Query.LinqOperatorProvider.<_TrackEntities>d__0`2.MoveNext() in D:\k\EntityFramework\src\EntityFramework.Core\Query\LinqOperatorProvider.cs:line 29
   at Microsoft.Data.Entity.Query.EntityQueryExecutor.EnumerableExceptionInterceptor`1.EnumeratorExceptionInterceptor.MoveNext() in D:\k\EntityFramework\src\EntityFramework.Core\Query\EntityQueryExecutor.cs:line 185
   at System.Collections.Generic.List`1..ctor(IEnumerable`1 collection)
   at System.Linq.Enumerable.ToList[TSource](IEnumerable`1 source)
   at IncludeBug_IndexOutsideBounds.Program.Main(String[] args) in D:\Projects\IncludeBug_IndexOutsideBounds\IncludeBug_IndexOutsideBounds\Program.cs:line 80

**Comment by maumar** [comment]
this still repros after multi-level include change

**Comment by maumar** [comment]
most likely same root cause as #1487

**Comment by maumar** [comment]
fixed in 617ad0aaa964949e4a620ef5a73fa4bbe2f12374
